File: paleo_workbench/ui/pages/table_preview_widget.py
from __future__ import annotations

from PySide6.QtCore import QAbstractTableModel, QItemSelectionModel, QModelIndex, Qt
from PySide6.QtGui import QBrush, QColor, QFont, QKeySequence
from PySide6.QtWidgets import (
    QApplication,
    QHeaderView,
    QSizePolicy,
    QTableView,
)

from paleo_workbench.ui import tokens

# Defensive payload bound only: the virtualized model removed the per-cell
# widget allocation that made 50k cells the freeze threshold (#1039), so the
# cap now merely guards against pathological parser output far beyond the
# settings maximum (2,000 rows × 200 cols = 400k cells). It is NOT used to
# hide reasonably sized tables from virtualized rendering.
MAX_PREVIEW_CELLS = 1_000_000

_DEPTH_HEADERS = frozenset({"DEPT", "DEPTH", "深度"})

_MONO = QFont("Cascadia Code", 9)
_MONO_BOLD = QFont("Cascadia Code", 9)
_MONO_BOLD.setBold(True)

# Highlight brushes resolve from the ACTIVE theme palette (#1047): fixed
# light-token brushes left dark themes with unreadable tints.
_THEMED_BRUSHES: dict[str, dict[str, QBrush]] = {}


def _themed_brushes() -> dict[str, QBrush]:
    from paleo_workbench.ui.theme import theme_manager

    theme = theme_manager.current_theme.value
    cached = _THEMED_BRUSHES.get(theme)
    if cached is None:
        palette = tokens.palette_for(theme)
        cached = {
            "depth_fg": QBrush(QColor(palette["PRIMARY"])),
            "depth_bg": QBrush(QColor(palette["BG_SELECTION"])),
            "tag_fg": QBrush(QColor(palette["TEAL"])),
            "tag_bg": QBrush(QColor(palette["BG_SEARCH"])),
            "unit_fg": QBrush(QColor(palette["TEXT_SECONDARY"])),
            "nan_fg": QBrush(QColor(palette["PRIMARY_DISABLED"])),
        }
        _THEMED_BRUSHES[theme] = cached
    return cached

_ALIGN_RIGHT = Qt.AlignmentFlag.AlignRight | Qt.AlignmentFlag.AlignVCenter
_ALIGN_CENTER = Qt.AlignmentFlag.AlignCenter | Qt.AlignmentFlag.AlignVCenter


class _CellRef:
    """Minimal read-only stand-in for QTableWidgetItem (legacy accessors)."""

    __slots__ = ("_text",)

    def __init__(self, text: str, _alignment) -> None:
        self._text = text

    def text(self) -> str:
        return self._text

# Column auto-fit only samples this many leading rows: resizeColumnsToContents
# walks every row, which reintroduces an O(rows × columns) pass on load.
_AUTO_FIT_SAMPLE_ROWS = 400


def _is_number(val: str) -> bool:
    if val == "NaN":
        return True
    try:
        float(val)
        return True
    except ValueError:
        return False


class TablePreviewModel(QAbstractTableModel):
    """Virtual model over detached row tuples (#1039).

    Cell text and formatting are computed on demand in ``data()`` for the
    visible viewport only — loading a 100,000-row preview allocates no
    per-cell widget or item, and scrolling never touches off-screen rows.
    """

    def __init__(self, parent=None) -> None:
        super().__init__(parent)
        self._headers: tuple[str, ...] = ()
        self._rows: tuple[tuple[str, ...], ...] = ()
        self._depth_column: int = -1
        self._is_curve_def: bool = False

    # -- population ---------------------------------------------------------

    def set_table(
        self,
        headers: tuple[str, ...],
        rows: tuple[tuple[str, ...], ...],
    ) -> None:
        self.beginResetModel()
        self._headers = tuple(str(h) for h in headers)
        self._rows = rows
        self._depth_column = -1
        for index, header in enumerate(self._headers):
            if header.upper() in _DEPTH_HEADERS:
                self._depth_column = index
                break
        self._is_curve_def = (
            len(self._headers) >= 3 and self._headers[0] in ("曲线", "Mnemonic")
        )
        self.endResetModel()

    # -- Qt model contract --------------------------------------------------

    def rowCount(self, parent: QModelIndex | None = None) -> int:  # noqa: N802
        if parent is not None and parent.isValid():
            return 0
        return len(self._rows)

    def columnCount(self, parent: QModelIndex | None = None) -> int:  # noqa: N802
        if parent is not None and parent.isValid():
            return 0
        return len(self._headers)

    def headerData(self, section, orientation, role=Qt.ItemDataRole.DisplayRole):  # noqa: N802
        if role != Qt.ItemDataRole.DisplayRole:
            return None
        if orientation == Qt.Orientation.Horizontal and 0 <= section < len(self._headers):
            return self._headers[section]
        if orientation == Qt.Orientation.Vertical and 0 <= section < len(self._rows):
            return str(section + 1)
        return None

    def data(self, index, role=Qt.ItemDataRole.DisplayRole):  # noqa: N802
        if not index.isValid():
            return None
        row = index.row()
        column = index.column()
        if not (0 <= row < len(self._rows) and 0 <= column < len(self._headers)):
            return None
        raw = self._rows[row][column] if column < len(self._rows[row]) else None
        val_str = str(raw).strip() if raw is not None else ""

        if role == Qt.ItemDataRole.DisplayRole:
            return val_str

        brushes = _themed_brushes()

        # 1. Depth column (DEPT / DEPTH / 深度)
        if column == self._depth_column:
            if role == Qt.ItemDataRole.FontRole:
                return _MONO_BOLD
            if role == Qt.ItemDataRole.ForegroundRole:
                return brushes["depth_fg"]
            if role == Qt.ItemDataRole.BackgroundRole:
                return brushes["depth_bg"]
            if role == Qt.ItemDataRole.TextAlignmentRole:
                return int(_ALIGN_RIGHT)

        # 2. Curve mnemonic tag formatting in curve definition table
        elif self._is_curve_def and column == 0:
            if role == Qt.ItemDataRole.FontRole:
                return _MONO_BOLD
            if role == Qt.ItemDataRole.ForegroundRole:
                return brushes["tag_fg"]
            if role == Qt.ItemDataRole.BackgroundRole:
                return brushes["tag_bg"]
            if role == Qt.ItemDataRole.TextAlignmentRole:
                return int(_ALIGN_CENTER)

        # 3. Unit column formatting
        elif self._is_curve_def and column == 1:
            if role == Qt.ItemDataRole.FontRole:
                return _MONO
            if role == Qt.ItemDataRole.ForegroundRole:
                return brushes["unit_fg"]
            if role == Qt.ItemDataRole.TextAlignmentRole:
                return int(_ALIGN_CENTER)

        # 4. Numeric curve data formatting
        elif _is_number(val_str):
            if role == Qt.ItemDataRole.FontRole:
                return _MONO
            if role == Qt.ItemDataRole.TextAlignmentRole:
                return int(_ALIGN_RIGHT)
            if role == Qt.ItemDataRole.ForegroundRole and val_str == "NaN":
                return brushes["nan_fg"]

        return None

    # -- programmatic access --------------------------------------------------

    def row_text(self, row: int) -> list[str]:
        """Formatted cell strings of one row (copy/export path)."""
        if not 0 <= row < len(self._rows):
            return []
        source = self._rows[row]
        return [
            (str(source[c]).strip() if c < len(source) and source[c] is not None else "")
            for c in range(len(self._headers))
        ]

    @property
    def headers(self) -> tuple[str, ...]:
        return self._headers


class TablePreviewWidget(QTableView):
    """Virtualized table preview: QTableView + TablePreviewModel (#1039).

    The legacy QTableWidget materialized one QTableWidgetItem per cell on the
    GUI thread (up to MAX_PREVIEW_CELLS), freezing the window on large
    previews. The model computes display data lazily for the viewport, so
    100,000 rows open without per-cell allocation.
    """

    def __init__(self, parent=None):
        super().__init__(parent)
        self._model = TablePreviewModel(self)
        self.setModel(self._model)
        self.setEditTriggers(QTableView.EditTrigger.NoEditTriggers)
        self.setAlternatingRowColors(True)
        self.setShowGrid(True)
        self.setSelectionBehavior(QTableView.SelectionBehavior.SelectItems)
        self.auto_fit_columns = True
        self.truncated = False
        self.truncation_message = ""

        # Alternating-row tint comes from the global themed sheet
        # (tokens.build_qss provides alternate-background-color for
        # QTableView across every palette — a widget-level light tint here
        # broke dark/high-contrast themes, #1047 review).
        self.setSizePolicy(QSizePolicy.Policy.Expanding, QSizePolicy.Policy.Expanding)
        self.horizontalHeader().setStretchLastSection(True)
        self.verticalHeader().setDefaultSectionSize(28)

    def apply_settings(self, settings) -> None:
        font = self.font()
        font.setPointSize(settings.font_size)
        self.setFont(font)
        self.auto_fit_columns = settings.auto_fit_columns
        mode = (
            QHeaderView.ResizeMode.ResizeToContents
            if self.auto_fit_columns
            else QHeaderView.ResizeMode.Interactive
        )
        self.horizontalHeader().setSectionResizeMode(mode)

    # -- QTableWidget-compatible accessors used by host panels -------------

    def rowCount(self) -> int:  # noqa: N802
        return self._model.rowCount()

    def columnCount(self) -> int:  # noqa: N802
        return self._model.columnCount()

    def rowHeight(self, row: int) -> int:  # noqa: N802
        return self.verticalHeader().sectionSize(row)

    def columnWidth(self, column: int) -> int:  # noqa: N802
        return self.horizontalHeader().sectionSize(column)

    def item(self, row: int, column: int):  # noqa: N802
        """Read-only on-demand cell reference (legacy accessor).

        Nothing is allocated at load or scroll time; the lightweight wrapper
        only exists for code that explicitly asks for a cell — the eager
        materialization that froze the GUI is gone (#1039).
        """
        index = self._model.index(row, column)
        if not index.isValid():
            return None
        return _CellRef(
            str(index.data(Qt.ItemDataRole.DisplayRole) or ""),
            index.data(Qt.ItemDataRole.TextAlignmentRole),
        )

    def horizontalHeaderItem(self, column: int):  # noqa: N802
        if not 0 <= column < len(self._model.headers):
            return None
        return _CellRef(self._model.headers[column], None)

    def setRangeSelected(self, table_range, select: bool) -> None:  # noqa: N802
        """Legacy QTableWidget selection entry (range-based).

        Maps onto the view's QItemSelectionModel so copy/selection semantics
        survive the virtualization unchanged (#1039).
        """
        if not select:
            return
        from PySide6.QtCore import QItemSelection

        top_left = self._model.index(table_range.topRow(), table_range.leftColumn())
        bottom_right = self._model.index(
            table_range.bottomRow(), table_range.rightColumn()
        )
        selection_model = self.selectionModel()
        if selection_model is None:
            return
        selection_model.select(
            QItemSelection(top_left, bottom_right),
            QItemSelectionModel.SelectionFlag.Select
            | QItemSelectionModel.SelectionFlag.Current,
        )

    def load_table(
        self,
        headers: tuple[str, ...],
        rows: tuple[tuple[str, ...], ...],
    ) -> None:
        self.truncated = False
        self.truncation_message = ""
        n_cols = len(headers)
        visible_rows = rows
        if n_cols > 0 and len(rows) * n_cols > MAX_PREVIEW_CELLS:
            keep = max(1, MAX_PREVIEW_CELLS // n_cols)
            visible_rows = rows[:keep]
            self.truncated = True
            self.truncation_message = (
                f"表格预览已截断：显示 {keep}/{len(rows)} 行"
                f"（上限 {MAX_PREVIEW_CELLS} 单元格）"
            )
        if self.truncated:
            self.setToolTip(self.truncation_message)
            self.setStatusTip(self.truncation_message)
        else:
            self.setToolTip("")
            self.setStatusTip("")
        self._model.set_table(tuple(headers), tuple(visible_rows))

        if self.auto_fit_columns:
            hdr = self.horizontalHeader()
            n_cols = self._model.columnCount()
            # Fit from a bounded SAMPLE of leading rows: Qt's
            # resizeColumnsToContents walks every row, which would reintroduce
            # the O(rows × columns) load-time pass this widget removed (#1039).
            hdr.setSectionResizeMode(QHeaderView.ResizeMode.Interactive)
            hdr.setStretchLastSection(False)
            sample_rows = [
                self._model.row_text(r)
                for r in range(min(self._model.rowCount(), _AUTO_FIT_SAMPLE_ROWS))
            ]
            metrics = self.fontMetrics()
            for col in range(n_cols):
                width = metrics.horizontalAdvance(self._model.headers[col]) + 24
                for cells in sample_rows:
                    if col < len(cells):
                        width = max(width, metrics.horizontalAdvance(cells[col]) + 24)
                hdr.resizeSection(col, max(width + 16, 75))
            hdr.setStretchLastSection(True)

    def copy_all(self) -> str:
        """返回当前显示表格的 TSV（含表头），仅复制已截断后的可见行。"""
        model = self._model
        n_cols = model.columnCount()
        if n_cols == 0:
            return ""
        headers = list(model.headers)
        lines: list[str] = ["\t".join(headers)]
        for r in range(model.rowCount()):
            lines.append("\t".join(model.row_text(r)))
        # 无数据行时仅返回表头
        if model.rowCount() == 0:
            return "\t".join(headers)
        return "\n".join(lines)

    def keyPressEvent(self, event) -> None:  # noqa: N802
        is_copy = False
        try:
            if event.matches(QKeySequence.StandardKey.Copy):
                is_copy = True
        except Exception:
            pass
        if not is_copy and event.key() == Qt.Key.Key_C and bool(
            event.modifiers() & Qt.KeyboardModifier.ControlModifier
        ):
            is_copy = True
        if is_copy:
            indexes = self.selectedIndexes()
            if not indexes:
                event.accept()
                return
            by_row: dict[int, list[tuple[int, str]]] = {}
            for index in indexes:
                by_row.setdefault(index.row(), []).append(
                    (index.column(), str(index.data()) or "")
                )
            lines: list[str] = []
            for row in sorted(by_row):
                cells = sorted(by_row[row])
                lines.append("\t".join(text for _, text in cells))
            text = "\n".join(lines)
            clipboard = QApplication.clipboard()
            if clipboard is not None:
                clipboard.setText(text)
            event.accept()
            return
        super().keyPressEvent(event)

File: libs/ui_pages_mapedit/src/table_preview_widget.cpp
#include "pwb/ui_pages_mapedit/table_preview_widget.hpp"

#include "pwb/ui_shell/style_registry.hpp"

#include <QAbstractTableModel>
#include <QApplication>
#include <QClipboard>
#include <QFontMetrics>
#include <QHeaderView>
#include <QItemSelection>
#include <QItemSelectionModel>
#include <QKeyEvent>
#include <QKeySequence>
#include <QSizePolicy>

namespace pwb::ui_pages_mapedit {
namespace {

// Defensive payload bound only (Python MAX_PREVIEW_CELLS): the virtualized
// model removed the per-cell widget allocation, so the cap merely guards
// against pathological parser output.
constexpr int kMaxPreviewCells = 1'000'000;
// Column auto-fit only samples this many leading rows (Python
// _AUTO_FIT_SAMPLE_ROWS): resizeColumnsToContents walks every row.
constexpr int kAutoFitSampleRows = 400;

bool is_depth_header(const QString& header) {
    const QString up = header.toUpper();
    return up == QLatin1String("DEPT") || up == QLatin1String("DEPTH") ||
           header == QLatin1String("深度");
}

const QFont& mono_font() {
    static const QFont font(QStringLiteral("Cascadia Code"), 9);
    return font;
}

const QFont& mono_bold_font() {
    static QFont font = [] {
        QFont f(QStringLiteral("Cascadia Code"), 9);
        f.setBold(true);
        return f;
    }();
    return font;
}

// Python ``float(val)`` accepts underscores between digits plus
// inf/infinity/nan; QString::toDouble covers the same numeric grammar once
// underscores are validated and removed.
bool is_number(const QString& val) {
    if (val == QLatin1String("NaN")) {
        return true;
    }
    QString s = val;
    for (int i = 0; i < s.size(); ++i) {
        if (s[i] == u'_') {
            const bool between_digits =
                i > 0 && i + 1 < s.size() && s[i - 1].isDigit() &&
                s[i + 1].isDigit();
            if (!between_digits) {
                return false;
            }
        }
    }
    s.remove(u'_');
    bool ok = false;
    s.toDouble(&ok);
    return ok;
}

// Highlight brushes resolve from the ACTIVE theme palette (#1047 parity):
// fixed light-token brushes left dark themes with unreadable tints.
struct ThemedBrushes {
    QBrush depth_fg;
    QBrush depth_bg;
    QBrush tag_fg;
    QBrush tag_bg;
    QBrush unit_fg;
    QBrush nan_fg;
};

QString pal(const std::map<std::string, std::string>& p, const char* key,
            const char* fallback) {
    const auto it = p.find(key);
    const QString v = it != p.end() ? QString::fromStdString(it->second)
                                    : QString::fromLatin1(fallback);
    return v;
}

ThemedBrushes themed_brushes() {
    const auto& p = ui_shell::style_palette();
    return {QBrush(QColor(pal(p, "PRIMARY", "#0b5563"))),
            QBrush(QColor(pal(p, "BG_SELECTION", "#d8ebef"))),
            QBrush(QColor(pal(p, "TEAL", "#0f766e"))),
            QBrush(QColor(pal(p, "BG_SEARCH", "#edf1f4"))),
            QBrush(QColor(pal(p, "TEXT_SECONDARY", "#53616c"))),
            QBrush(QColor(pal(p, "PRIMARY_DISABLED", "#8c99a3")))};
}

constexpr int kAlignRight =
    static_cast<int>(Qt::AlignRight | Qt::AlignVCenter);
constexpr int kAlignCenter =
    static_cast<int>(Qt::AlignCenter | Qt::AlignVCenter);

}  // namespace

// TablePreviewModel — headers + string rows, display computed on demand
// (the Python virtualization contract: no per-cell allocation at load).
class TablePreviewModel : public QAbstractTableModel {
    Q_OBJECT
public:
    explicit TablePreviewModel(QObject* parent = nullptr)
        : QAbstractTableModel(parent) {}

    void set_table(const QStringList& headers,
                   const std::vector<QStringList>& rows) {
        beginResetModel();
        headers_ = headers;
        rows_ = rows;
        depth_column_ = -1;
        for (int i = 0; i < headers_.size(); ++i) {
            if (is_depth_header(headers_[i])) {
                depth_column_ = i;
                break;
            }
        }
        is_curve_def_ = headers_.size() >= 3 &&
                        (headers_[0] == QLatin1String("曲线") ||
                         headers_[0] == QLatin1String("Mnemonic"));
        endResetModel();
    }

    int rowCount(const QModelIndex& parent = QModelIndex()) const override {
        return parent.isValid() ? 0 : static_cast<int>(rows_.size());
    }
    int columnCount(const QModelIndex& parent = QModelIndex()) const override {
        return parent.isValid() ? 0 : headers_.size();
    }
    QVariant headerData(int section, Qt::Orientation orientation,
                        int role = Qt::DisplayRole) const override {
        if (role != Qt::DisplayRole) {
            return {};
        }
        if (orientation == Qt::Horizontal && section >= 0 &&
            section < headers_.size()) {
            return headers_[section];
        }
        if (orientation == Qt::Vertical && section >= 0 &&
            section < static_cast<int>(rows_.size())) {
            return QString::number(section + 1);
        }
        return {};
    }

    QVariant data(const QModelIndex& index,
                  int role = Qt::DisplayRole) const override {
        if (!index.isValid()) {
            return {};
        }
        const int row = index.row();
        const int column = index.column();
        if (!(row >= 0 && row < static_cast<int>(rows_.size()) &&
              column >= 0 && column < headers_.size())) {
            return {};
        }
        const auto& source = rows_[row];
        const QString val_str =
            column < source.size() ? source[column].trimmed() : QString();

        if (role == Qt::DisplayRole) {
            return val_str;
        }

        const ThemedBrushes brushes = themed_brushes();

        // 1. Depth column (DEPT / DEPTH / 深度)
        if (column == depth_column_) {
            if (role == Qt::FontRole) {
                return mono_bold_font();
            }
            if (role == Qt::ForegroundRole) {
                return brushes.depth_fg;
            }
            if (role == Qt::BackgroundRole) {
                return brushes.depth_bg;
            }
            if (role == Qt::TextAlignmentRole) {
                return kAlignRight;
            }
        }
        // 2. Curve mnemonic tag formatting in curve definition table
        else if (is_curve_def_ && column == 0) {
            if (role == Qt::FontRole) {
                return mono_bold_font();
            }
            if (role == Qt::ForegroundRole) {
                return brushes.tag_fg;
            }
            if (role == Qt::BackgroundRole) {
                return brushes.tag_bg;
            }
            if (role == Qt::TextAlignmentRole) {
                return kAlignCenter;
            }
        }
        // 3. Unit column formatting
        else if (is_curve_def_ && column == 1) {
            if (role == Qt::FontRole) {
                return mono_font();
            }
            if (role == Qt::ForegroundRole) {
                return brushes.unit_fg;
            }
            if (role == Qt::TextAlignmentRole) {
                return kAlignCenter;
            }
        }
        // 4. Numeric curve data formatting
        else if (is_number(val_str)) {
            if (role == Qt::FontRole) {
                return mono_font();
            }
            if (role == Qt::TextAlignmentRole) {
                return kAlignRight;
            }
            if (role == Qt::ForegroundRole &&
                val_str == QLatin1String("NaN")) {
                return brushes.nan_fg;
            }
        }
        return {};
    }

    // Formatted cell strings of one row (copy/export path).
    QStringList row_text(int row) const {
        if (row < 0 || row >= static_cast<int>(rows_.size())) {
            return {};
        }
        const auto& source = rows_[row];
        QStringList out;
        out.reserve(headers_.size());
        for (int c = 0; c < headers_.size(); ++c) {
            out.append(c < source.size() ? source[c].trimmed() : QString());
        }
        return out;
    }

    const QStringList& headers() const { return headers_; }

private:
    QStringList headers_;
    std::vector<QStringList> rows_;
    int depth_column_ = -1;
    bool is_curve_def_ = false;
};

TablePreviewWidget::TablePreviewWidget(QWidget* parent)
    : QTableView(parent), model_(new TablePreviewModel(this)) {
    setModel(model_);
    setEditTriggers(QTableView::NoEditTriggers);
    setAlternatingRowColors(true);
    setShowGrid(true);
    setSelectionBehavior(QTableView::SelectItems);
    setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    horizontalHeader()->setStretchLastSection(true);
    verticalHeader()->setDefaultSectionSize(28);
}

void TablePreviewWidget::apply_settings(int font_size, bool auto_fit) {
    QFont f = font();
    f.setPointSize(font_size);
    setFont(f);
    auto_fit_columns = auto_fit;
    horizontalHeader()->setSectionResizeMode(
        auto_fit_columns ? QHeaderView::ResizeToContents
                         : QHeaderView::Interactive);
}

void TablePreviewWidget::load_table(
    const QStringList& headers, const std::vector<QStringList>& rows) {
    truncated = false;
    truncation_message.clear();
    const int n_cols = headers.size();
    const std::vector<QStringList>* visible_rows = &rows;
    std::vector<QStringList> kept;
    if (n_cols > 0 &&
        static_cast<long long>(rows.size()) * n_cols > kMaxPreviewCells) {
        const int keep =
            std::max(1, kMaxPreviewCells / n_cols);
        kept.assign(rows.begin(), rows.begin() + std::min<int>(
                                     keep, static_cast<int>(rows.size())));
        visible_rows = &kept;
        truncated = true;
        truncation_message = QStringLiteral(
            "表格预览已截断：显示 %1/%2 行（上限 %3 单元格）")
            .arg(keep)
            .arg(rows.size())
            .arg(kMaxPreviewCells);
    }
    if (truncated) {
        setToolTip(truncation_message);
        setStatusTip(truncation_message);
    } else {
        setToolTip(QString());
        setStatusTip(QString());
    }
    model_->set_table(headers, *visible_rows);

    if (auto_fit_columns) {
        auto* hdr = horizontalHeader();
        const int cols = model_->columnCount();
        // Fit from a bounded SAMPLE of leading rows: resizeColumnsToContents
        // walks every row — the O(rows × columns) pass this widget removed.
        hdr->setSectionResizeMode(QHeaderView::Interactive);
        hdr->setStretchLastSection(false);
        const int sample =
            std::min(model_->rowCount(), kAutoFitSampleRows);
        std::vector<QStringList> sample_rows;
        sample_rows.reserve(sample);
        for (int r = 0; r < sample; ++r) {
            sample_rows.push_back(model_->row_text(r));
        }
        const QFontMetrics metrics = fontMetrics();
        for (int col = 0; col < cols; ++col) {
            int width =
                metrics.horizontalAdvance(model_->headers().at(col)) + 24;
            for (const auto& cells : sample_rows) {
                if (col < cells.size()) {
                    width = std::max(
                        width, metrics.horizontalAdvance(cells[col]) + 24);
                }
            }
            hdr->resizeSection(col, std::max(width + 16, 75));
        }
        hdr->setStretchLastSection(true);
    }
}

QString TablePreviewWidget::copy_all() const {
    const int n_cols = model_->columnCount();
    if (n_cols == 0) {
        return QString();
    }
    QStringList lines;
    lines.append(model_->headers().join(QLatin1Char('\t')));
    for (int r = 0; r < model_->rowCount(); ++r) {
        lines.append(model_->row_text(r).join(QLatin1Char('\t')));
    }
    if (model_->rowCount() == 0) {
        return model_->headers().join(QLatin1Char('\t'));
    }
    return lines.join(QLatin1Char('\n'));
}

void TablePreviewWidget::keyPressEvent(QKeyEvent* event) {
    bool is_copy = event->matches(QKeySequence::Copy);
    if (!is_copy && event->key() == Qt::Key_C &&
        (event->modifiers() & Qt::ControlModifier)) {
        is_copy = true;
    }
    if (is_copy) {
        const QModelIndexList indexes = selectedIndexes();
        if (indexes.isEmpty()) {
            event->accept();
            return;
        }
        std::map<int, std::map<int, QString>> by_row;
        for (const QModelIndex& index : indexes) {
            by_row[index.row()][index.column()] =
                index.data().toString();
        }
        QStringList lines;
        for (const auto& [row, cells] : by_row) {
            QStringList texts;
            for (const auto& [col, text] : cells) {
                texts.append(text);
            }
            lines.append(texts.join(QLatin1Char('\t')));
        }
        if (QClipboard* clipboard = QApplication::clipboard()) {
            clipboard->setText(lines.join(QLatin1Char('\n')));
        }
        event->accept();
        return;
    }
    QTableView::keyPressEvent(event);
}

int TablePreviewWidget::rowCount() const {
    return model_->rowCount();
}

int TablePreviewWidget::columnCount() const {
    return model_->columnCount();
}

QString TablePreviewWidget::item_text(int row, int column) const {
    const QModelIndex index = model_->index(row, column);
    return index.isValid() ? index.data(Qt::DisplayRole).toString()
                           : QString();
}

QStringList TablePreviewWidget::row_text(int row) const {
    return model_->row_text(row);
}

QString TablePreviewWidget::horizontal_header_text(int column) const {
    if (column < 0 || column >= model_->headers().size()) {
        return QString();
    }
    return model_->headers()[column];
}

void TablePreviewWidget::set_range_selected(int top_row, int left_column,
                                            int bottom_row, int right_column,
                                            bool select) {
    if (!select) {
        return;
    }
    const QModelIndex top_left = model_->index(top_row, left_column);
    const QModelIndex bottom_right =
        model_->index(bottom_row, right_column);
    QItemSelectionModel* selection_model = selectionModel();
    if (selection_model == nullptr) {
        return;
    }
    selection_model->select(
        QItemSelection(top_left, bottom_right),
        QItemSelectionModel::Select | QItemSelectionModel::Current);
}

}  // namespace pwb::ui_pages_mapedit

#include "table_preview_widget.moc"
